contracts/HubPool.sol

```diff
@@ -405,6 +421,37 @@ contract HubPool is Testable, Lockable, MultiCaller, Ownable {
                 .rawValue;
     }
 
+    function _executeRelayerRefundOnTargetChain(MerkleLib.PoolRebalance memory poolRebalance, uint256 refundRequestId)
+        internal
+    {
+        AdapterInterface adapter = crossChainContracts[poolRebalance.chainId].adapter;
+        adapter.relayMessage(
+            address(adapter),
+            abi.encodeWithSignature("initializeRelayerRefund(address)", refundRequest.destinationDistributionRoot)
+        );
+    }
+
+    function _sendTokensToTargetChain(MerkleLib.PoolRebalance memory poolRebalance, uint256 refundRequestId) internal {
+        AdapterInterface adapter = crossChainContracts[poolRebalance.chainId].adapter;
+        for (uint32 i = 0; i < poolRebalance.tokenAddresses.length; i++) {
+            int256 amount = poolRebalance.netSendAmount[i];
+
+            // TODO: Checking the amount is greater than 0 is not sufficient. we need to build an external library that
+            // makes the decision on if there should be an L1->L2 token transfer. this should come in a later PR.
+            if (amount > 0) {
+                // Send the adapter all the tokens it needs to bridge. This should be refined later to remove the extra
+                // token transfer through the use of delegate call.
+                IERC20(poolRebalance.tokenAddresses[i]).safeTransfer(address(adapter), uint256(amount));
+                adapter.relayTokens(
```

Comment:
is there any way, or even an advantage to, potentially batching these calls?

Reply:
I dont think we can batch these easily as this is the actual action that sends tokens from L1->L2 so we'd need a way to batch these across the canonical bridge.